---
title: "FOS 2021 GENOME-WIDE ASSOCIATION ANALYSIS PRACTICAL"
author: "Insert your name here"
email: "Insert your email here"
date: "Insert the date here"

output: html_notebook
---

# 1. INTRODUCTION
Genome-wide association study (GWAS) is a study design used to identify association between genetic variants and diseases or quantitative traits.^[Tam, Vivian, et al. "Benefits and limitations of genome-wide association studies." Nature Reviews Genetics 20.8 (2019): 467-484.] Typically, GWAS is a hypothesis-free approach that involves looking for a single base-pair change in the DNA sequence (single nucleotide polymorphisms or SNPs) that occur with high frequency in the human genome.^[Genomes Project Consortium (2010) A map of human genome variation from population-scale sequencing. Nature 467: 1061â€“1073 .] SNPs are said to be "associated" with disease/trait if certain SNPs are more commonly found in people with a disease/trait compared to those without the disease/trait. SNPs are typically used as markers of a genomic region, with most of them having a minimal impact on biological systems.^[Bush, William S., and Jason H. Moore. "Chapter 11: Genome-wide association studies." PLoS computational biology 8.12 (2012): e1002822.]

In this exercise, you will learn how to perform the genome-wide
association analysis using rvtests. Rvtests is a comprehensive tool
developed to support genetic association analysis. Details about rvtests can be found on <http://zhanxw.github.io/rvtests/>. The phenotype of interest in todays' practical is a metabolically related quaternary ammonium compound measured in blood, betaine. Genetic data includes genotypes imputed with Haplotype Reference Consortium imputation panel.
<br>

# 2. INSTRUCTIONS
For this practical there are 2 assessment files:
FOS_GWASpractical_20211124.html
FOS_GWASpractical_20211124.RMD

You will use the html file to follow the practical, while you use the RMD file to run code that is required to answer the questions.
Open the .html file on your browser by clicking over "FOS_GWASpractical_20211124.html" in the Files window of RStudio (bottom right) and pressing "View in Web Browser".

Please, add your name, email and date on top of this file where it says "Insert your name here", "Insert your email here" and "Insert the date here" and reply to the questions with your answers wherever you see the **A.**.
At the end of the practical, please save this RMD file and upload this RMD file with the answers to Turnitin on Brightspace.

Load the libraries required for the practicals:
```{r libraries, echo=F ,results=F}
library (data.table)  
library(gap)  
```
**Note:** Some of the packages might not have been installed yet. If receiving an error run: *install.packages ("gap")* and try again.

# 3. GET FAMILIAR WITH THE DATA

### 3.1 PHENOTYPE DATA
```{r phenotypes}
pheno <- read.table ("Mydata.pheno", header=TRUE, stringsAsFactors=F)
dim (pheno)
head (pheno)

covar <- read.table ("Mydata.covar", header=TRUE, stringsAsFactors=F)
dim (covar)
head (covar)
```

**EXCERCISES**

**Q1.** How many individuals are in the dataset? What percentage is female?

```{r eval=FALSE}
table (pheno$Sex)
```

**A1.** Please type your answer here.


**Q2.** Inspect variable betaine. Create descriptive statistics for this variable (mean, standard deviation, median, minimum, maximum) and generate histogram. What do the descriptive statistics look like?

```{r eval=FALSE}
summary (pheno$Betaine) 
sd (pheno$Betaine,na.rm=TRUE) 
hist (pheno$Betaine, main = paste("Histogram of betaine - original variable")) 
```

**A2.** Please type your answer here.


**Q3.** There may be several issues with this betaine variable? Why is this the case (look for missing values - strange number)? Correct these issues by setting the strange numbers of variable to missing (hint: check coding of missing values for quantitative trait in rvtests).

```{r eval=FALSE}
pheno$Betaine[pheno$Betaine==999.99] <- NA
```

**A3.** Please type your answer here.


**Q4.** Check descriptive statistics and generate a histogram of variable betaine again. Inspect the output. Is the variable normally distributed? Should we apply normalization? How should we treat this variable?

```{r eval=FALSE}
hist (pheno$Betaine, main = paste("Histogram of betaine - corrected variable"))
```

**A4.** Please type your answer here.


**Q5.** Now normalize the betaine variable by inverse rank transformation. To create a new variable, you could use code below. Generate a histogram of normalized variable and inspect the normalized values.

```{r eval=FALSE}
pheno$Betaine_rank<- qnorm((rank(pheno$Betaine,na.last="keep")-0.5)/sum(!is.na(pheno$Betaine)))
hist (pheno$Betaine_rank, main = paste("Histogram of betaine - rank transformed")) 
```

**A5.** Please type your answer here.


### 3.2. GENOTYPE DATA
Open Mydata.vcf file in R and answer the questions.

```{r}
# Read header of the file
vcf_header <- readLines("Mydata.vcf")
head(vcf_header)
  
# Read the file
vcf <- fread ("Mydata.vcf")
head (vcf)

```

**EXCERCISES**

**Q6.** What are the sub-fields included in the FORMAT field in the
Mydata.vcf file? (hint: check header of the file)\
**A6.** Please type your answer here.

**Q7.** What is the ID name of first two samples listed in the Mydata.vcf file?\
**A7.** Please type your answer here.

**Q8.** What are the position of the first two variants listed in the VCF file? What are the reference and alternative bases for these variants?\
**A8.** Please type your answer here.

**Q9.** How are genotypes encoded in the VCF file? What are the allele values for reference and alternative alleles?\
**A9.** Please type your answer here.

**Q10.** What is the genotype of first person for the first variant?\
**A10.** Please type your answer here.

## 4. QUALITY CONTROL PRIOR GWAS
**EXCERCISES**

**Q11.** Could you think of some strategies for avoiding bias induced by population stratification?\
**A11.** Please type your answer here.

**Q12.** Are there any samples that represent ethnic outlines on mdsplot? How would you deal with those samples in the GWAS?\
**A12.** Please type your answer here.

**Q13.** Does rvtests deal with related sample and how (hint: check
webpage)?\
**A13.** Please type your answer here.

**Q14.** Would you consider returning findings of sex chromosomal anomalies to individual if discovered during the sex-checks? What is the rationale behind this?
**A14.** Please type your answer here.

## 5. ASSOCIATION ANALYSIS
**EXCERCISE**

**Q15.** Based on the coding of the SNP in Mydata.vcf, additive model will be tested.If you would like to run dominant or recessive model, how would you recode the genotypes?\
**A15.** Please type your answer here.

**Q16.** As indicated in the model, covariates are age and sex.Are these confounders in the classical sense? If not, why would you still want to adjust for confounders in GWAS?
**A16.** Please type your answer here.

## 6. INSPECTION OF OUTPUT FILES
```{r load results}
results <- read.table (gzfile("Myresults.txt.gz"),header=TRUE, stringsAsFactors=F)

results[1:5,1:16]
```

**EXCERCISES**

**Q17.** How many samples are analyzed for association? How many SNPs are included in Myresults.txt.gz file? Which chromosomes are these SNPs mapped to?

```{r results, eval=FALSE}
dim (results) 
table(results$CHROM)
```
**A17.** Please type your answer here.

**Q18.** How many genome-wide significant variants are in Myresults.txt.gz file?

```{r significant results,eval=FALSE}
dim(results[which(results$PVALUE<0.00000005),]) 
```
**A18.** Please type your answer here.

## 7. VISUALIZATION
### 7.1 QQ PLOT
```{r qqplot}
qq = function(pvector, ...) { if (!is.numeric(pvector)) stop("D'oh! P value vector is not numeric.") 
pvector <- pvector[!is.na(pvector) & pvector<1 & pvector>0] 
o = -log10(sort(pvector,decreasing=F)) 
e = -log10(ppoints(length(pvector) )) 
plot(e,o,pch=19,cex=1,xlab=expression(Expected~~-log[10](italic(p))),ylab=expression(Observed~~-log[10](italic(p))), xlim=c(0,max(e)),
ylim=c(0,max(o)), ...) 
abline(0,1,col="red") }

qq(results$PVALUE) 
```

**EXCERCISES**

**Q19.** Check the QQ plot. Does it look acceptable?\
**A19.** Please type your answer here.

**Q20.** A QQ plot made with the data of different study and different quantitative trait is displayed below.
![QQ plot](https://drive.google.com/uc?export=view&id=1HSHI01hFGrREAuPR6hbvzgqbTH33PkXG)
Is the quality of the GWAS acceptable according to this QQ plot? If not, what are the reasons?
**A20.** Please type your answer here.

**EXCERCISES**

**Q21.** Calculate the inflation factor. How do you interpret it?

```{r eval=FALSE}
results$SE <- 1/results$SQRT_V_STAT 
chi2 <- (results$ALT_EFFSIZE/results$SE)^2 
lam <- median(sort(chi2))/qchisq(0.5,1) 
```
**A21.** Please type your answer here.

### 7.2 MANHATTAN PLOT
```{r}
library (gap)
# select chromosome,position, and p-value name
results1<-results[,c("CHROM","POS","PVALUE")]  
# change names to work with the functions
names(results1) <- c("chr","pos","p")
# remove NA's
results1<-results1[!is.na(results1$p),] 
# order the file
results1 <- results1[order(results1[,1], results1[,2]),]

# if you have 22 chromosome change to: color <- rep(c("blue","red"),11) 
color <- rep(c("blue","red"),1) 
# for 22 chrom, change labels to: labels=1:22
ops <- mht.control(colors=color,yline=1.5,xline=1,labels=c(5,6),srt=270) 
mhtplot(results1[,c("chr","pos","p")],ops,xlab="chromosomes",ylab="-log p-value",srt=0) 
axis(2,at=1:round(max(-log10(results1$p))))
abline(h=5,col="grey",lty=1) 
abline(h=-log10(0.00000005),col="dark green",lty=1) 
rm(results1)
gc()
```

**EXCERCISES**

**Q22.** What is the p-value threshold of the dark green line?\
**A22.** Please type your answer here.

**Q23.** Is there any SNP genome widely significantly associated with betaine? If yes, where is it located?\
**A23.** Please type your answer here.

```{r eval=FALSE}
head(results[order(results$PVALUE),])
```

**Q24.** Does top hit make sense?
**A24.** Please type your answer here.

### 7.3 LOCUS ZOOM PLOT
**EXERCISES**

**Q25.** How does the locus zoom plot look like for the top SNP? Is there linkage disequilibrium between the top SNP and other SNP of the same region?\
**A25.** Please type your answer here.

**Q26.** Go back to the results file and check the imputation quality, allele frequency and HWE of this top SNP. Is this SNP reliable?
**A26.** Please type your answer here.

```{r eval=FALSE}
results[which(results$ID=="6:171048850_T_C"),]
```

## 8. POST GWAS QUALITY CONTROL
**EXERCISES**

**Q27.**  How many SNPs with MAF < 0.01 are in your dataset?\
```{r eval=FALSE}
dim(results[which(results$AF<0.01 | results$AF>0.99),])
```
**A27.** Please type your answer here.

**Q28.** What is AF of 5:12180_T_G SNP? How many carriers of minor allele are present in your data?\
```{r eval=FALSE}
results[which(results$ID=="5:12180_T_G"),]
```
**A28.** Please type your answer here.

**Q29.** What are the lowest p-value for HWE test, the lowest imputation quality and call rate in Myresults.txt.gz?\
```{r eval=FALSE}
head(results[order(results$HWE_PVALUE),])
head(results[order(results$CALL_RATE),])
head(results[order(results$Rsq),])
```
**A29.** Please type your answer here.

**Q30.** How many SNPs are out of HWE if you use a p-value threshold of 0.000001? Which factors can cause deviation from HWE?\
```{r eval=FALSE}
dim(results[order(results$HWE_PVALUE<0.000001),])
```
**A30.** Please type your answer here.

**Q31.** How many SNPs do you filter out based on MAF < 0.01, call rate of 95%, HWE p-value of 0.000001 or imputation quality of 0.3?\
**A31.** Please type your answer here.

**Q32.** Create a file listing all SNPs that passed above mentioned SNP-specific level filters. How many genome wide significant variants are there in your meta-analysis after performing the QC?\
**A32.** Please type your answer here.

**Q33.** Create Manhattan, QQ and locus zoom plots using the QC-ed result file. The QC'ed file has been uploaded to the locus zoom website. Go to https://my.locuszoom.org/gwas/571690/?token=55852dc8d8514c6481e76807dbf41d6f to generate regional plots for your top findings. Do you notice any difference between plots generated based on non-QC-ed and QC-ed summary statistic results?\
**A33.** Please type your answer here.

**Q34.** Where are your top SNPs located?\
**A34.** Please type your answer here.

## 9. ANNOTATION
**EXERCISES**

**Q35.** Do the results make any sense given that the trait is circulating level of betaine? (hint: What is the function of the top genes?)\
**A35.** Please type your answer here.

**Q36.** Which pathways are these genes involved in?\
**A36.** Please type your answer here.

**Q37.** Did any other GWAS find the similar answer as you did?\
**A37.** Please type your answer here.


## 10. VALIDATION STUDIES
**EXERCISES**

**Q38.** Would you use similar of different population for replication and why?\
**A38.** Please type your answer here.
**Q39.** Which SNPs would you select for replication and why?\
**A39.** Please type your answer here.
**Q40.** Would you involve meta-analysis of the original summary? statistic results and the results from validation study and why?\
**A40.** Please type your answer here.

## 11. SESSIONINFO

```{r}
sessionInfo()
```